### ISSUE TITLE
Support for M1/ARM64

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#1162]
@Igazine reported some issues for installing Moonshine on an M1/AMD64 Mac:

![image](https://user-images.githubusercontent.com/24532731/206248705-6b80d7cc-5616-46ff-8dd6-6b1205c582e1.png)

Also note that there are some issues with the JDK:  https://github.com/Moonshine-IDE/Moonshine-SDK-Installer/issues/98

**Comment by rat-moonshine** [comment]
Based on the Apple [documents](https://support.apple.com/en-in/HT211861) "Which app needs Rosetta?", it seems Moonshine may not need Rosetta as it falls under "Universal" category.

**Comment by Igazine** [comment]
## Notes and observations regarding host CPU architecture compatibility (Part 1)

### Rosetta

Apple's relatively new Silicon chips are based on the ARM64 architecture, which means apps that are built for x86_64 (Intel) architecture can no longer work on the new Macs. To ease the transition, Apple introduced Rosetta 2 in 2020 (with Big Sur), which is a translation/application compatibility layer between ARM64 and x86_64 architectures. So in order to run x86_64 based apps on Silicon Macs, one has to have Rosetta installed on macOS. Any time you try to install or launch an x86_64 based application (including our extensively used development frameworks, runtimes, IDEs etc.), macOS shows a dialog stating that you need to install Rosetta. This has to be done only once, and once it's installed it'll never warn you again, Intel apps will just run through Rosetta without the user even noticing it.

Some of you might think that you don't have Rosetta installed, but in reality it's very easy to overlook, as it displays a very common looking installation dialog with a blue "Install" button (please see the attached image at the top) often when you're trying to install something x86_64 based, so probably many of you already have it installed. Some of the app you're trying to install (eg. Parallels, XCode) even install it without notifying the user.

### How to determine if Rosetta is installed

In Terminal, enter

`pgrep oahd`

This will print the process id of the Rosetta daemon out (Rosetta is internally called OAH). If it doesn't print anything, Rosetta is not installed. Alternatively

`pgrep -q oahd && echo Yes || echo No`

will print Yes or No according to whether Rosetta is installed or not.

### How to uninstall Rosetta

In order to test on a pure ARM64 environment, Rosetta shouldn't be installed. To uninstall Rosetta, [follow these instructions](https://iboysoft.com/news/uninstall-rosetta-2.html). **But be warned, the process requires System Integrity Protection (SIP) to be disabled for a short time.**

**Comment by Igazine** [comment]
## Notes and observations regarding host CPU architecture compatibility (Part 2)

### How to determine if an executable is ARM or x64 based?

The `file {path-to-file}` command shows information about the given file. To see which architecture an executable supports, point it to the selected executable. To test Moonshine, enter the followings in Terminal:

```
file /Applications/Moonshine.app/Contents/MacOS/Moonshine
```

or

```
file /Applications/MoonshineDevelopment.app/Contents/MacOS/MoonshineDevelopment
```


It'll show that the main executable is an universal binary that supports both architectures. Now enter:

```
file /Applications/MoonshineDevelopment.app/Contents/Frameworks/SecurityScopeBookmark.framework/SecurityScopeBookmark
```

to see that the SecurityScopeBookmark extension is currently x86_64 only, which means it must be rebuilt (with possibly the latest Harman AIR) and bundled in the app. I tested Moonshine on an M1 Mac without Rosetta, and while the app runs, this extension fails to load and Moonshine throws an error in the console.

**Comment by Igazine** [comment]
## Notes and observations regarding host CPU architecture compatibility (Part 3)

### Installer (.pkg)

The installer itself (the pkg file) is x86_64 only, so it'll show the Rosetta dialog the first time you're trying to install Moonshine. According to our build.xml, the pkg is created with ADT. The reason it's an x86-only installer can either be a misconfiguration or an older version or AIR on our build servers. Whatever the reason is, this issue should be investigated.

This is the Distribution file from the pkg.

![image](https://user-images.githubusercontent.com/2218406/206979585-e021f8d1-ab02-4157-a7df-3cd0cdef9e6e.png)

**Comment by Igazine** [comment]
## Notes and observations regarding host CPU architecture compatibility (Part 4)

### Software Support

In order to achieve a fully ARM64 Moonshine, the bundled software needs to be ARM64-compatible as well. Currently the main Moonshine app only contains Node and Python modules, and JAR files, which should be running perfectly fine on ARM64 as long as the ARM64 versions of NodeJS, Python, and Java are installed. The previously mentioned SecurityScopeBookmark extension still needs to be recompiled, but otherwise I could not find anything bundled within the app that requires a different binary.

### MoonshineSDKInstaller

Most of the required/optional frameworks, runtimes etc. are being installed with MoonshineSDKInstaller. Needless to say that MSDKI and its installer also has to be ARM64-compatible.

```
file /Applications/MoonshineSDKInstaller.app/Contents/MacOS/Moonshine\ SDK\ Installer
```

This will show that the main executable of MSDKI support both architectures. The installer, however, is x86_64 only similarly to Moonshine's installer.

### Software installed by MoonshineSDKInstaller

All of the available installers in MSDKI must point to the appropriate ARM64 installers/binaries in order to achieve native ARM64 compatibility.

#### Architecture support matrix

| Software | ARM64 | Notes |
| -------- | -------- | -------- |
| Adobe AIR | No | Discontinued |
| Harman AIR | Yes | |
| Apache Royale | Yes | Node module |
| Haxe | No | Haxe can built from sources on ARM64 or installed with Homebrew, but there's no official ARM64 binary and installer |
| HXCPP | No | HXCPP is not compatible with ARM64 on the assembly level, so C++ apps cannot be created |
| Lime/OpenFL | No | Currently Lime/OpenFL is not supported on ARM64 natively |
| Neko | No | If Haxe is installed with Homebrew, ARM64 Neko will also be installed. There's no official ARM64 download |
| Java 8 | No | |
| Java 17 | Yes | |
| Apache Ant | Yes | Jar files and shell scripts |
| Maven | Yes | Jar files and shell scripts |
| Gradle | Yes | Jar files and shell scripts |
| Grails | Yes | Jar files and schell scripts |
| Subversion | Yes | Installed with MacPorts |
| Git | Yes | Can be installed with XCode, MacPorts, Homebrew etc. |
| NodeJS | Yes | |
| HCL Notes | No | |
| MacPorts | Yes | |
| Vagrant | No | Homebrew installation is cask-only, there's no official ARM64 build |

**Comment by Igazine** [comment]
## Notes and observations regarding host CPU architecture compatibility (Part 5)

### Moonshine's future

The bad news is that Haxe (more precisely HXCPP, the C++ compiler for Haxe) is currently not compatible with ARM64 on the assembly level. While Haxe (and Neko) can be built from sources or can be installed with Homebrew on ARM64, unfortunately Native/C++ applications cannot be created.

Some targets work perfectly fine (e.g. Java, Python, Haxe Interpreter etc.), I could even manage to create and bootstrap native Neko binaries, but creating C++ apps is currently not possible.

Lime also requires some libraries to be present for Lime/OpenFL apps, which are also x86_64-only currently.

```
file /usr/local/lib/haxe/lib/lime/8,0,0/ndll/Mac64/lime.ndll
```

This will show that the Mac64 lime.ndll is an x86_64 binary.

### Conclusion

Currently it's very difficult to use a Silicon Mac without Rosetta being installed, and not just because of Moonshine. There are many apps out there that require Rosetta, so sooner or later every Silicon Mac user will be forced to install it (or it will be installed in the background without the user even noticing it).

The question is how long will Apple support Rosetta. There's no official statement regarding the length of support, but judging from the previous architecture change, Rosetta might be with us for several years to come, and by the time it goes away, all of the required/used software will natively support ARM64.

**Comment by rat-moonshine** [comment]
> I tested Moonshine on an M1 Mac without Rosetta, and while the app runs, this extension fails to load and Moonshine throws an error in the console.

Okay. So Moonshine itself runs without Rosetta but it fails to work on SecurityBookmark. Thanks for pointing to this again.

Bamboo build system compiles on Harman AIR SDK, and possibly on not so old build. Re-building the SecurityBookmark with Harman files, could be troublesome, though. I need to revisit the documentation if available.

**Comment by Igazine** [comment]
I managed to rebuild the SecurityBookmarkExtension and the ANE as well with Harman AIR SDK v 33.1.1.929 to support both architectures. As far as I can tell (using the Ant buildfile and running Moonshine from an IDE) the extension works fine on Intel and M1 Macs (even with forced arm64 architecture, i.e. `arch -arm64 ./Moonshine.app/Contents/MacOS/Moonshine`)

@rat-moonshine  Please let me know if you need me to commit and push the changes in a new branch for you to test. Thank you

**Comment by rat-moonshine** [comment]
Thank you @Igazine . Hopefully the `SSB.ane` meets the required ARM support (`file SSB.ane`) that you discussed [here](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1162#issuecomment-1345976810).

Yes, I appreciate if I can give a test at my side, too. I'd suggest to commit the `SSB.ane` only to `master` for now, and once we finished with initial testing please commit the SSB project files as well.

Let me know when you commits the ANE.

**Comment by Igazine** [comment]
@rat-moonshine I just pushed the new ANE, please give the build a test if it finishes.

With Rosetta installed it'll require the following commands to be tested (after the installation of the new build).
First make sure the new executable supports both architectures:

```
file /Applications/MoonshineDevelopment.app/Contents/Frameworks/SecurityScopeBookmark.framework/SecurityScopeBookmark
```
This should print the support for both ARM64 and X86_64 support.

Then forcing the app to run in arm64-only mode (works on M1/M2 Macs only)
```
arch -arm64 /Applications/MoonshineDevelopment.app/Contents/MacOS/MoonshineDevelopment
```
The error mentioned in [this comment](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1162#issuecomment-1345976810) should not occur.

**Comment by Igazine** [comment]
Please also make sure that the extension itself works as expected on both architectures

**Comment by Igazine** [comment]
I've updated the ANE files, there was an additional directory and a file in the package that broke codesign.
Please try again. Thank you

**Comment by rat-moonshine** [comment]
Unfortunately, the error continues to trigger after the update @Igazine .



```
13-Dec-2022 03:11:53 | [exec] DEPLOY/MoonshineDevelopment.app: unsealed contents present in the root directory of an embedded framework
-- | --
13-Dec-2022 03:11:53 | [exec] In subcomponent: /Users/mmb/bamboo-agent-home/xml-data/build-dir/MS-MGOSX-JOB1/Moonshine-IDE/ide/MoonshineDESKTOPevolved/build/DEPLOY/MoonshineDevelopment.app/Contents/Frameworks/SecurityScopeBookmark.framework
13-Dec-2022 03:11:53 |  
13-Dec-2022 03:11:53 | BUILD FAILED
```

**Comment by Igazine** [comment]
I pushed one more change in the latest commit, the ANE is now created with proper symlinks in it (the previous versions contained an extra `./` at the beginning of the links and it looks like it broke codesign.

Please try again one more time. Thank you

**Comment by rat-moonshine** [comment]
The last update (symlink) passed the signing procedure. I shall give some tests with newer builds once it completes.

**Comment by rat-moonshine** [comment]
A macOS build is ready with the updated ANE file build by @Igazine and can be tested with build >=#2,952.

Following are initial test results.

Mac Arc | Sandbox | Opens Previously Accessed File/Folder | Console Error on ARM | Test Notes
---------|-----------|-------------------------|------------|------
ARM | No | Yes | No | -
ARM | Yes | Yes | No | -
ARM | Yes | No | No | If previously accessed scope removed by going into `File > Access Manager`, previously accessed file/folder unable to access, as expected. An error/alert shows in that case.
X64 | Pending | Pending | Pending | -

**Comment by JoelProminic** [comment]
With these updates, will Moonshine prompt the user to install Rosetta if it isn't installed already?  Or will this be skipped because the main Moonshine application doesn't require Rosetta.

I found some [documentation](https://developer.apple.com/documentation/virtualization/running_intel_binaries_in_linux_vms_with_rosetta) that may help with detecting and prompting for Rosetta.

To test, we'll need a way to uninstall Rosetta or otherwise simulate an environment where Rosetta is missing.

**Comment by Igazine** [comment]
You can also simulate a "missing" Rosetta with the `arch` command. E.g.:
```
arch -arm64 /Applications/MoonshineDevelopment.app/Contents/MacOS/MoonshineDevelopment
```
launches MoonshineDevelopment (or any other executable you point it to) in arm64-only mode (ignoring Rosetta)

The rest are described in my comments above.

**Comment by JoelProminic** [comment]
From our discussion, the only remaining issue is the installer, which I moved to a separate issue (#1169)
